src/vdaf/mastic/hh_sketch.rs

```diff
@@ -0,0 +1,359 @@
+        let bucket = u32::from_be_bytes(<[u8; 4]>::try_from(&hash[..4]).unwrap());
+        let sign = hash[5] & 1 == 1;
+        Ok((sign, bucket))
+    }
+}
+
+impl Type for HeavyHittersSketch {
+    type Measurement = Vec<bool>;
+    type AggregateResult = Option<Vec<bool>>;
+
+    fn encode_measurement(&self, measurement: &Vec<bool>) -> Result<Vec<Field64>, FlpError> {
+        let (sign, _bucket) = self.sign_and_bucket(measurement)?;
+
+        // [RZCGP24], Figure 5:
+        //
+        // The first field element, denoted `beta`, is computed by hashing the measurement: if
+        // `beta == 0`, then we compliment the bits of the input before encoding.
+        //
+        // For each bit of the input, if `bit == 1`, then we encoded it as `1`; otherwise we encode
+        // it as `-1`.
+        //
+        // TODO Rewrite this to avoid branching on the value of `sign` or `bit`.
+        Ok(once(if sign {
+            Field64::one()
+        } else {
+            -Field64::one()
+        })
+        .chain(measurement.iter().map(|bit| {
+            if bit ^ !sign {
+                Field64::one()
+            } else {
+                -Field64::one()
+            }
+        }))
+        .collect())
+    }
+
+    fn truncate(&self, input: Vec<Field64>) -> Result<Vec<Field64>, FlpError> {
+        Ok(input)
+    }
+
+    fn decode_result(
+        &self,
+        data: &[Field64],
+        num_measurements: usize,
+    ) -> Result<Option<Vec<bool>>, FlpError> {
+        if num_measurements < self.threshold {
+            return Ok(None);
+        }
+
+        let m = Field64::modulus() - u64::try_from(num_measurements).unwrap();
+
+        // [RZCGP24], Figure 5: Round the data by mapping each value larger than `n` to `0` and
+        // every other value to `1`.
+        let sign = (1..m).contains(&u64::from(data[0]));
+        let heavy_hitter = data[1..]
+            .iter()
+            .copied()
+            .map(|x| (1..m).contains(&u64::from(x)) ^ !sign)
+            .collect();
+        Ok(Some(heavy_hitter))
+    }
+
+    fn output_len(&self) -> usize {
+        self.input_len()
+    }
+}
+
+impl Flp for HeavyHittersSketch {
+    type Field = Field64;
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<Field64>>> {
+        // p(x) = (x+1)(x-1)
+        let p = vec![-Field64::one(), Field64::zero(), Field64::one()];
+        vec![Box::new(PolyEval::new(p, 1 + usize::from(self.num_bits)))]
+    }
+
+    fn num_gadgets(&self) -> usize {
+        1
+    }
+
+    fn valid(
+        &self,
+        gadgets: &mut Vec<Box<dyn Gadget<Field64>>>,
+        input: &[Field64],
+        joint_rand: &[Field64],
+        _num_shares: usize,
+    ) -> Result<Vec<Field64>, FlpError> {
+        self.valid_call_check(input, joint_rand)?;
+
+        // Check that each input is either a `1` or a `-1`.
+        input
+            .iter()
+            .copied()
+            .map(|x| gadgets[0].call(&[x]))
+            .collect()
```

Comment:
A typical input length will probably be something like `num_bits == 256`. I wonder if we'd be better off (in terms of proof size and computation) with `ParallelSum` here.

That would mean we would need joint randomness and thus a larger field (`Field128`). I'm not sure if a larger field would pay off.